templates/android/library/src/main/java/io/appwrite/services/Realtime.kt.twig

```diff
@@ -0,0 +1,157 @@
+import {{ sdk.namespace | caseDot }}.exceptions.{{ spec.title | caseUcfirst }}Exception
+import {{ sdk.namespace | caseDot }}.extensions.forEachAsync
+import {{ sdk.namespace | caseDot }}.extensions.fromJson
+import {{ sdk.namespace | caseDot }}.extensions.jsonCast
+import {{ sdk.namespace | caseDot }}.models.RealtimeCodes
+
+import {{ sdk.namespace | caseDot }}.models.RealtimeMessage
+import {{ sdk.namespace | caseDot }}.models.RealtimeSubscription
+import kotlinx.coroutines.*
+import kotlinx.coroutines.Dispatchers.IO
+import okhttp3.*
+import okhttp3.internal.concurrent.TaskRunner
+import okhttp3.internal.ws.RealWebSocket
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+class Realtime(client: Client) : Service(client), CoroutineScope {
+
+    private val job = Job()
+
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.Main + job
+
+    private companion object {
+        private const val DEBOUNCE_MILLIS = 1L
+
+        private var socket: RealWebSocket? = null
+        private var channelCallbacks = mutableMapOf<String, MutableCollection<(Map<*, *>) -> Unit>>()
+        private var errorCallbacks = mutableSetOf<({{ spec.title | caseUcfirst }}Exception) -> Unit>()
+
+        private var subCallDepth = 0
+    }
+
+    private fun createSocket() {
+        val queryParamBuilder = StringBuilder()
+            .append("project=${client.config["project"]}")
+
+        channelCallbacks.keys.forEach {
+            queryParamBuilder
+                .append("&channels[]=$it")
+        }
+
+        val request = Request.Builder()
+            .url("${client.endPointRealtime}/realtime?$queryParamBuilder")
+            .build()
+
+        if (socket != null) {
+            closeSocket()
+        }
+
+        socket = RealWebSocket(
+            taskRunner = TaskRunner.INSTANCE,
+            originalRequest = request,
+            listener = {{ spec.title | caseUcfirst }}WebSocketListener(),
+            random = Random(),
+            pingIntervalMillis = client.http.pingIntervalMillis.toLong(),
+            extensions = null,
+            minimumDeflateSize = client.http.minWebSocketMessageToCompress
+        )
+
+        socket!!.connect(client.http)
+    }
+
+    private fun closeSocket() {
+        socket?.close(RealtimeCodes.POLICY_VIOLATION.value, null)
+    }
+
+    fun subscribe(
+        vararg channels: String,
+        callback: (Map<*, *>) -> Unit
+    ) : RealtimeSubscription {
+        channels.forEach {
+            if (!channelCallbacks.containsKey(it)) {
+                channelCallbacks[it] = mutableListOf(callback)
+                return@forEach
+            }
+            channelCallbacks[it]?.add(callback)
+        }
+
+        launch {
+            subCallDepth++
+            delay(DEBOUNCE_MILLIS)
+            if (subCallDepth == 1) {
+                createSocket()
+            }
+            subCallDepth--
+        }
+
+        return RealtimeSubscription { unsubscribe(*channels) }
```

Comment:
Do we return the unsubscribe method right away or an object containing the method?

Reply:
This returns a `RealtimeSubscription` object that has a `close` function, there's a couple of reasons I did it this way:
- It implements [`java.io.Closeable`](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html), which is the Kotlin/Android way to represent closable resources (and allows usage with try-with-resoruces or .use() constructs)
- We could add other info to the subscription model later without it being a breaking SDK change